### PR TITLE
PP-726: Systemd PBS service startup times out if there are tens of thousands of jobs present

### PR DESCRIPTION
**Pull request by sid2364**
#### Issue-ID

* **[PP-726](https://pbspro.atlassian.net/browse/PP-726)**

#### Problem description
* PBS service times out if there are tens of thousands of jobs present

#### Cause / Analysis
* Timeout value was set to 5 minutes.

#### Solution description
* Changed timeout value to infinity.

#### Checklist:

- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [x] I have added  **manual test(s) to the Jira ticket and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to the Jira ticket** as evidence of testing/verification.

__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

### CODE REVIEW

File: src/cmds/scripts/pbs.service.in [hunk 1]
```diff
@@ -55,7 +55,7 @@ DefaultDependencies=true
 [Service]
 Type=forking
 Restart=no
-TimeoutSec=5min
+TimeoutSec=0
```
Comment (mike0042) [inline src/cmds/scripts/pbs.service.in]: I assume this means that the timeout is disabled. Perhaps a larger value would make more sense? I'm not certain that waiting indefinitely is the correct behavior.

Comment (sid2364) [inline src/cmds/scripts/pbs.service.in]: We don't know of a problem with an infinite timeout, which is why I changed it to that. All daemons handle cases where coming up is blocked by network or external environment; they "continue in background".

Comment (mike0042) [inline src/cmds/scripts/pbs.service.in]: If we understand the ramifications of setting this to zero, then I'm fine with it. My concern is that it was initially set to five minutes, ostensibly for some purpose. Why would any service choose to limit its startup time if it doesn't have to? What action(s) does systemd take if a timeout occurs? It appears that systemd is becoming an obstacle in more ways than one as of late.

Comment (sid2364) [inline src/cmds/scripts/pbs.service.in]: Systemd terminates the start up process on reaching timeout. Setting it to infinity should not be a problem, at least not as far as I can tell.

**Comment by prakashcv13** [comment]
I suggest we use TimeoutStartSec instead of TimeoutSec as the latter affects both start and stop time and here we are concerned only with start time.

**Comment by sid2364** [comment]
@prakashcv13 In older versions of systemd, it is "0" instead of "infinity" to disable timeout logic. Also, it's backward compatible. :)
[Source](https://github.com/systemd/systemd/blob/master/NEWS#L1361).

Will change to TimeoutStartSec instead.

**Comment by prakashcv13** [comment]
I would suggest to use the latest and not rely on backward compatibility.

**Comment by sid2364** [comment]
Using "infinity" will not work with older systems, and would in fact cause PBS to always fail on startup due to timeout since they don't recognise "infinity" as a valid value for TimeoutSec.

**Comment by prakashcv13** [comment]
hmmm... right!

**Comment by sid2364** [comment]
@prakashcv13 Have changed it to TimeoutStartSec.

**Comment by prakashcv13** [comment]
@sid2364, as discussed offline, please work on automating the scenario using PTL..

**Comment by arungrover** [comment]
As I understand the issue is about server taking too long to come up because it is busy recovering jobs in its foreground process.
When we know that server already goes in background when it isn't able to connect to database initially while coming up then why don't we just go to background always and then load the jobs?
I may be wrong here but, If we make it infinite then wouldn't it affect system startups where server will be busy recovering jobs for minutes and startup will be stuck?

**Comment by prakashcv13** [comment]
pep8 check has failed...

**Comment by sid2364** [comment]
Have added PTL test case.

File: test/tests/performance/pbs_startup_timeout.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+        with open(jobscript, "wb") as js:
+            js.write("#!/bin/sh")
+            js.write("/bin/sleep 20")
+            js.truncate(1024 * 1024 * 5)
+
+        # Submit 1000 jobs
+        job = Job()
+        job.set_execargs(jobscript)
+        for _ in range(1000):
```
Comment (arungrover) [inline test/tests/performance/pbs_startup_timeout.py]: 1000 isn't really a big number, do you think that server will take more than 300 seconds to come up when there are 1000 jobs in db.
I believe this test script will pass with/without your change.
I don't really know how you can test this in a PTL script (maybe you don't need a PTL script for this pull)
Besides "self.server.isUp()" is just a "ps -e" of pbs_server process which will always give you a positive result. you might want to check for "systemd --show-status"

Comment (sid2364) [inline test/tests/performance/pbs_startup_timeout.py]: Actually we decided that we can't really verify this fix using PTL (I forgot to take the PTL off this PR). What we could do instead is put a sleep in the init script and that would "simulate" loading jobs into memory and we could test that the TimeoutStart is functional.

Will take the PTL off this PR.

**Comment by sid2364** [comment]
@arungrover Updated the PR.

**Comment by arungrover** [comment]
@sid2364  I sign-off. Once you have another sign-off please rebase the branch
It's interesting that when you removed the file and pushed it, it didn't trigger any CI build :)

**Comment by sid2364** [comment]
I did a git reset --hard to revert the last commit, so maybe it didn't consider this as a "new" commit? CI build should start when I rebase though.

**Comment by arungrover** [comment]
yes, that explains it

File: test/tests/performance/pbs_startup_timeout.py (hunk 2 above)
Comment (iestockdale) [inline test/tests/performance/pbs_startup_timeout.py]: The bug description says that PBS "fails rather messily with only a few tens of thousands of jobs".  What is missing from PTL leading to an inability to write this test?

**Comment by JonShelley** [comment]
This was seen at a site that was using GPFS for PBS_HOME which is a slow file system for our application. Using different file systems that handle small reads and writes better would require significantly more jobs to display the same behavior

Thanks,

Jon Shelley
PBS Works – Director
jshelley@altair.com
PBS Support – pbssupport@altair.com

**Comment by sid2364** [comment]
@PBSPro/pbspro-maintainers Rebased.

**Comment by arungrover** [comment]
@sid2364 please rebase it again

**Comment by sid2364** [comment]
@arungrover Rebased it again.

**Comment by subhasisb** [comment]
I got it Arun - will merge whenever Sid tell me he is updated again.

**Comment by sid2364** [comment]
@subhasisb Have rebased it again.

**Pull request merged**